File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/view/adapter/list/GroupListAdapter.java
package app.android.ttp.mikazuki.yoshinani.view.adapter.list;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import app.android.ttp.mikazuki.yoshinani.R;
import app.android.ttp.mikazuki.yoshinani.binding.BindableString;
import app.android.ttp.mikazuki.yoshinani.model.GroupModel;
import app.android.ttp.mikazuki.yoshinani.model.TotalModel;
import app.android.ttp.mikazuki.yoshinani.utils.TextUtils;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author haijimakazuki
 */
public class GroupListAdapter extends ArrayAdapter<GroupListAdapter.GroupListItem> {
    private final Context mContext;
    private final LayoutInflater mLayoutInflater;

    public GroupListAdapter(Context context,
                            List<GroupListItem> groups) {
        super(context, 0, groups);
        mContext = context;
        mLayoutInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    public static List<GroupListItem> createGroupListItems(List<GroupModel> activeGroups,
                                                           List<GroupModel> invitedGroups,
                                                           Map<Integer, TotalModel> totals) {
        List<GroupListItem> items = new ArrayList<>();
        for (GroupModel group : activeGroups) {
            items.add(new GroupListItem(group, totals.get(group.getId()), true));
        }
        for (GroupModel group : invitedGroups) {
            items.add(new GroupListItem(group, null, false));
        }
        return items;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;

        if (convertView == null) {
            convertView = mLayoutInflater.inflate(R.layout.list_group, parent, false);
            holder = new ViewHolder(convertView);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        GroupListItem group = getItem(position);
        holder.name.setText(group.getName().get());
        if (group.mIsActive) {
            holder.name.setTextColor(mContext.getResources().getColor(R.color.grey800));
            holder.amount.setText(TextUtils.wrapCurrency(group.getTotalResult()));
            holder.amount.setTextColor(mContext.getResources().getColor(group.getTotalResult() >= 0 ? R.color.theme600 : R.color.red800));
        } else {
            holder.name.setTextColor(mContext.getResources().getColor(R.color.grey400));
            holder.amount.setText("招待中");
            holder.amount.setTextColor(mContext.getResources().getColor(R.color.grey400));
        }
        return convertView;
    }

    static class ViewHolder {
        @BindView(R.id.name)
        TextView name;
        @BindView(R.id.amount)
        TextView amount;

        public ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }

    public static class GroupListItem {

        private GroupModel mGroup;
        private int mTotalResult;
        private boolean mIsActive;

        public GroupListItem(GroupModel group, TotalModel total, boolean isActive) {
            mGroup = group;
            mTotalResult = total != null ? total.getResult() : 0;
            mIsActive = isActive;
        }

        public BindableString getName() {
            return mGroup.getName();
        }

        public int getTotalResult() {
            return mTotalResult;
        }

    }
}

File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/view/adapter/list/GroupListAdapter.kt
package app.android.ttp.mikazuki.yoshinani.view.adapter.list

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView

import java.util.ArrayList

import app.android.ttp.mikazuki.yoshinani.R
import app.android.ttp.mikazuki.yoshinani.binding.BindableString
import app.android.ttp.mikazuki.yoshinani.model.GroupModel
import app.android.ttp.mikazuki.yoshinani.model.TotalModel
import app.android.ttp.mikazuki.yoshinani.utils.TextUtils
import butterknife.BindView
import butterknife.ButterKnife

/**
 * @author haijimakazuki
 */
class GroupListAdapter(private val mContext: Context,
                       groups: List<GroupListItem>) : ArrayAdapter<GroupListAdapter.GroupListItem>(mContext, 0, groups) {
    private val mLayoutInflater: LayoutInflater

    init {
        mLayoutInflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        val holder: ViewHolder

        if (convertView == null) {
            convertView = mLayoutInflater.inflate(R.layout.list_group, parent, false)
            holder = ViewHolder(convertView)
            convertView!!.tag = holder
        } else {
            holder = convertView.tag as ViewHolder
        }

        val group = getItem(position)
        holder.name!!.text = group!!.name.get()
        if (group.mIsActive) {
            holder.name!!.setTextColor(mContext.resources.getColor(R.color.grey800))
            holder.amount!!.text = TextUtils.wrapCurrency(group.totalResult.toDouble())
            holder.amount!!.setTextColor(mContext.resources.getColor(if (group.totalResult >= 0) R.color.theme600 else R.color.red800))
        } else {
            holder.name!!.setTextColor(mContext.resources.getColor(R.color.grey400))
            holder.amount!!.text = "招待中"
            holder.amount!!.setTextColor(mContext.resources.getColor(R.color.grey400))
        }
        return convertView
    }

    internal class ViewHolder(view: View) {
        @BindView(R.id.name)
        var name: TextView? = null
        @BindView(R.id.amount)
        var amount: TextView? = null

        init {
            ButterKnife.bind(this, view)
        }
    }

    class GroupListItem(private val mGroup: GroupModel, total: TotalModel?, private val mIsActive: Boolean) {
        val totalResult: Int

        val name: BindableString
            get() = mGroup.name

        init {
            totalResult = total?.result ?: 0
        }

    }

    companion object {

        fun createGroupListItems(activeGroups: List<GroupModel>,
                                 invitedGroups: List<GroupModel>,
                                 totals: Map<Int, TotalModel>): List<GroupListItem> {
            val items = ArrayList<GroupListItem>()
            for (group in activeGroups) {
                items.add(GroupListItem(group, totals[group.id], true))
            }
            for (group in invitedGroups) {
                items.add(GroupListItem(group, null, false))
            }
            return items
        }
    }
}